core/rawdb/taiko_l1_origin.go

```diff
@@ -29,17 +29,42 @@ func l1OriginKey(blockID *big.Int) []byte {
 
 // L1Origin represents a L1Origin of a L2 block.
 type L1Origin struct {
+	BlockID       *big.Int       `json:"blockID" gencodec:"required"`
+	L2BlockHash   common.Hash    `json:"l2BlockHash"`
+	L1BlockHeight *big.Int       `json:"l1BlockHeight"`
+	L1BlockHash   common.Hash    `json:"l1BlockHash"`
+	BatchID       *big.Int       `json:"batchID"`
+	EndOfBlock    bool           `json:"endOfBlock"`
+	EndOfPreconf  bool           `json:"endOfPreconf"`
+	Preconfer     common.Address `json:"preconfer"`
+}
+
+// L1OriginLegacy represents the legacy L1Origin structure of a L2 block.
+type L1OriginLegacy struct {
 	BlockID       *big.Int    `json:"blockID" gencodec:"required"`
 	L2BlockHash   common.Hash `json:"l2BlockHash"`
-	L1BlockHeight *big.Int    `json:"l1BlockHeight" gencodec:"required"`
-	L1BlockHash   common.Hash `json:"l1BlockHash" gencodec:"required"`
+	L1BlockHeight *big.Int    `json:"l1BlockHeight"`
+	L1BlockHash   common.Hash `json:"l1BlockHash"`
 }
 
 type l1OriginMarshaling struct {
 	BlockID       *math.HexOrDecimal256
 	L1BlockHeight *math.HexOrDecimal256
 }
 
+// IsSoftblock returns true if the L1Origin is a softblock.
+func (l *L1Origin) IsSoftblock() bool {
+	if l.BatchID == nil {
+		return true
+	}
+
+	if l.BatchID.Cmp(common.Big0) == 0 && l.Preconfer == (common.Address{}) {
```

Comment:
`BatchID == 0` also belongs to softblock.

Reply:
1) so we also check `l.Preconfer` 2) When u save this `l1Origin` to db, `0` will be the default value when u load them next time, no matter if its `nil` or not when u saved it.